### ISSUE TITLE
Glitchy Mode 5 when HD Mode 7 scaled at 1x and then upscaled at a later stage

### ISSUE THREAD

**Issue by Azrael256** [DerKoun/bsnes-hd#106]
I am talking here about the libretro implementation, which allows to upscale at the libretro/GUI stage, with shaders for pixel art upscaling. This allows HD Mode 7 in bsnes-hd to be set at 1x, offering great performances, most of the HD-Mode7 goodness, and the ability to use shaders such as hqx or nnedi. It works reasonably well in most situations, but seem to lead to glitchy screens in Mode 5.

![Here's an example of the glitch in action with HD Mode 7 Scale set to 1x](https://user-images.githubusercontent.com/92335732/136858584-49f74935-de2d-4562-b275-d8360a05b394.png)

![And here's how it looks like with HD Mode 7 Scale set to Disabled](https://user-images.githubusercontent.com/92335732/136858614-f5e3a3f6-f712-4ab8-88f9-f39f0bf156fd.png)

I may be wrong but I think HD Mode 7 should not be active at all in Mode 5. Unless I missed something, I do not think it yields any significant benefit and it seems to be mostly a cause for trouble. Otherwise, it does seem to be buggy at 1x and may need fixing. The problem disappears at 2x scaling or more.

**Comment by Azrael256** [comment]
How did it work before 10.2? Because it seemed to be working fine in betas 9/10. Was this, paradoxically, a bug that was fixed from 10.2 onward? 

Also, would it be possible and sensible to apply shaders before the upscaling stage, allowing for upscaling filters to work correctly whatever the final render resolution, and also make it much less costly in terms of resources... but I guess some shaders, e.g. CRT emulation shaders, would not necessarily work great that way?

**Comment by DerKoun** [comment]
Comparing the old and new images I think there may indeed be a bug here, maybe drawing pixels twice, horizontally. I'll look into that. Hopefully I just have to exclude some changes from 1x specifically. Thanks for the additional images and research.

**Comment by Azrael256** [comment]
Aaaaah that explains it. I did notice that upscalers cease to work in Mode 7 scenes even in normal bsnes. So it's a trade-off between allowing for preprocessing features and allowing for resolution-dependant post-processing features. I guess the trade-off used by bsnes-hd makes sense, using pixel-art upscaling shaders is a bit of an edge-case and rendering in HD from the get go creates an opportunity for pretty cool features.

**Comment by DerKoun** [comment]
Yes, some types of enhancements are mutually exclusive in nature.

There already is limited downscaling, for Mode 7 only. I have tried it for the completed HD image, but there are still issues with that code that I need to fix. (Also, technical note: It should not be "nearest neighbor", as that would eliminate the gained detail. It must blend n by n pixel squares into one pixel each, n being the scale factor.)
